From ba60a7a028329c0c6bba6e9e329a00b2cff96b65 Mon Sep 17 00:00:00 2001
From: cnauroth <cnauroth@apache.org>
Date: Tue, 7 Apr 2015 13:33:11 -0700
Subject: [PATCH 1253/2848] HDFS-8038. PBImageDelimitedTextWriter#getEntry
 output HDFS path in platform-specific format.
 Contributed by Xiaoyu Yao.

(cherry picked from commit 672ed462965ea7da10ce3df48c02e2a3bd13f0ae)
(cherry picked from commit 950dc1b1995191327e6b919ee17a87e59b749264)
(cherry picked from commit 5021f52f7bf65c3002e1c1767060aa728847c533)

Change-Id: I0d7059019bdcb8becac197e65040d3bd191899ad
---
 .../PBImageDelimitedTextWriter.java                |    8 +++++---
 .../offlineImageViewer/PBImageTextWriter.java      |    7 +++++--
 2 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/PBImageDelimitedTextWriter.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/PBImageDelimitedTextWriter.java
index 493d505..8fab34c 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/PBImageDelimitedTextWriter.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/PBImageDelimitedTextWriter.java
@@ -17,13 +17,13 @@
  */
 package org.apache.hadoop.hdfs.tools.offlineImageViewer;
 
+import org.apache.hadoop.fs.Path;
 import org.apache.hadoop.fs.permission.PermissionStatus;
 import org.apache.hadoop.hdfs.server.namenode.FsImageProto.INodeSection.INode;
 import org.apache.hadoop.hdfs.server.namenode.FsImageProto.INodeSection.INodeDirectory;
 import org.apache.hadoop.hdfs.server.namenode.FsImageProto.INodeSection.INodeFile;
 import org.apache.hadoop.hdfs.server.namenode.FsImageProto.INodeSection.INodeSymlink;
 
-import java.io.File;
 import java.io.IOException;
 import java.io.PrintStream;
 import java.text.SimpleDateFormat;
@@ -79,8 +79,10 @@ private void append(StringBuffer buffer, String field) {
   @Override
   public String getEntry(String parent, INode inode) {
     StringBuffer buffer = new StringBuffer();
-    String path = new File(parent, inode.getName().toStringUtf8()).toString();
-    buffer.append(path);
+    String inodeName = inode.getName().toStringUtf8();
+    Path path = new Path(parent.isEmpty() ? "/" : parent,
+      inodeName.isEmpty() ? "/" : inodeName);
+    buffer.append(path.toString());
     PermissionStatus p = null;
     boolean isDir = false;
     boolean hasAcl = false;
diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/PBImageTextWriter.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/PBImageTextWriter.java
index 0d2d6c3..1415913 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/PBImageTextWriter.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/tools/offlineImageViewer/PBImageTextWriter.java
@@ -23,6 +23,7 @@
 import com.google.common.io.LimitInputStream;
 import org.apache.commons.io.FileUtils;
 import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
 import org.apache.hadoop.fs.permission.PermissionStatus;
 import org.apache.hadoop.hdfs.server.namenode.FSImageFormatPBINode;
 import org.apache.hadoop.hdfs.server.namenode.FSImageFormatProtobuf;
@@ -146,7 +147,8 @@ private String getPath() {
           return "/";
         }
         if (this.path == null) {
-          this.path = new File(parent.getPath(), name).toString();
+          this.path = new Path(parent.getPath(), name.isEmpty() ? "/" : name).
+              toString();
           this.name = null;
         }
         return this.path;
@@ -375,7 +377,8 @@ public String getParentPath(long inode) throws IOException {
         }
         String parentName = toString(bytes);
         String parentPath =
-            new File(getParentPath(parent), parentName).toString();
+            new Path(getParentPath(parent),
+                parentName.isEmpty()? "/" : parentName).toString();
         dirPathCache.put(parent, parentPath);
       }
       return dirPathCache.get(parent);
-- 
1.7.9.5

